File: src/git.js
import simpleGit from 'simple-git';

const git = simpleGit();

export async function isGitRepository() {
  try {
    await git.revparse(['--git-dir']);
    return true;
  } catch (error) {
    return false;
  }
}

export async function getDiff(commitRange) {
  try {
    // Handle different input formats:
    // "HEAD~3" -> diff from HEAD~3 to HEAD
    // "HEAD~5..HEAD~2" -> diff between two commits
    let diffCommand;

    if (commitRange.includes('..')) {
      diffCommand = commitRange;
    } else {
      // Single commit reference - compare it to HEAD
      diffCommand = `${commitRange}..HEAD`;
    }

    // Ignore whitespace changes by default
    const diff = await git.diff([
      diffCommand,
      '--unified=3',
      '-w', // Ignore whitespace
      '--ignore-blank-lines'
    ]);
    return diff;
  } catch (error) {
    throw new Error(`Failed to get diff: ${error.message}`);
  }
}

export async function getDiffStats(commitRange) {
  try {
    let diffCommand;

    if (commitRange.includes('..')) {
      diffCommand = commitRange;
    } else {
      diffCommand = `${commitRange}..HEAD`;
    }

    // Get diff statistics
    const stats = await git.diff([
      diffCommand,
      '--stat',
      '-w',
      '--ignore-blank-lines'
    ]);

    // Parse the stats
    const lines = stats.trim().split('\n');
    const summaryLine = lines[lines.length - 1];

    // Extract numbers from summary like: "2 files changed, 50 insertions(+), 4 deletions(-)"
    const filesMatch = summaryLine.match(/(\d+) files? changed/);
    const insertionsMatch = summaryLine.match(/(\d+) insertions?/);
    const deletionsMatch = summaryLine.match(/(\d+) deletions?/);

    return {
      filesChanged: filesMatch ? parseInt(filesMatch[1]) : 0,
      insertions: insertionsMatch ? parseInt(insertionsMatch[1]) : 0,
      deletions: deletionsMatch ? parseInt(deletionsMatch[1]) : 0,
      raw: stats
    };
  } catch (error) {
    throw new Error(`Failed to get diff stats: ${error.message}`);
  }
}

export async function getCommitInfo(commitRef) {
  try {
    const log = await git.log([commitRef, '-1']);
    return log.latest;
  } catch (error) {
    throw new Error(`Failed to get commit info: ${error.message}`);
  }
}

export async function getCommitRange(range) {
  try {
    let fromRef, toRef;

    if (range.includes('..')) {
      [fromRef, toRef] = range.split('..');
    } else {
      fromRef = range;
      toRef = 'HEAD';
    }

    const log = await git.log([`${fromRef}..${toRef}`]);
    return log.all;
  } catch (error) {
    throw new Error(`Failed to get commit range: ${error.message}`);
  }
}

File: src/git.rs
use anyhow::{Context, Result};
use git2::Repository;
use std::process::Command;

pub struct DiffStats {
    pub files_changed: usize,
    pub insertions: usize,
    pub deletions: usize,
}

pub fn is_git_repo() -> Result<bool> {
    Ok(Repository::discover(".").is_ok())
}

pub fn get_current_branch() -> Result<String> {
    let repo = Repository::discover(".")?;
    let head = repo.head()?;

    if let Some(name) = head.shorthand() {
        Ok(name.to_string())
    } else {
        Ok("HEAD".to_string())
    }
}

pub fn get_diff(range: &str) -> Result<String> {
    let range = parse_range(range);

    let output = Command::new("git")
        .args(&[
            "diff",
            &range,
            "--unified=3",
            "-w",  // Ignore whitespace
            "--ignore-blank-lines",
        ])
        .output()
        .context("Failed to execute git diff")?;

    if !output.status.success() {
        anyhow::bail!("git diff failed: {}", String::from_utf8_lossy(&output.stderr));
    }

    Ok(String::from_utf8(output.stdout)?)
}

pub fn get_diff_stats(range: &str) -> Result<DiffStats> {
    let range = parse_range(range);

    let output = Command::new("git")
        .args(&[
            "diff",
            &range,
            "--stat",
            "-w",
            "--ignore-blank-lines",
        ])
        .output()
        .context("Failed to execute git diff --stat")?;

    if !output.status.success() {
        anyhow::bail!("git diff --stat failed: {}", String::from_utf8_lossy(&output.stderr));
    }

    let stats_text = String::from_utf8(output.stdout)?;
    parse_stats(&stats_text)
}

fn parse_range(range: &str) -> String {
    if range.contains("..") {
        range.to_string()
    } else {
        format!("{}..HEAD", range)
    }
}

fn parse_stats(text: &str) -> Result<DiffStats> {
    let lines: Vec<&str> = text.trim().split('\n').collect();
    if lines.is_empty() {
        return Ok(DiffStats {
            files_changed: 0,
            insertions: 0,
            deletions: 0,
        });
    }

    let summary_line = lines.last().unwrap_or(&"");

    let files_changed = extract_number(summary_line, "file");
    let insertions = extract_number(summary_line, "insertion");
    let deletions = extract_number(summary_line, "deletion");

    Ok(DiffStats {
        files_changed,
        insertions,
        deletions,
    })
}

fn extract_number(text: &str, pattern: &str) -> usize {
    for part in text.split(',') {
        if part.contains(pattern) {
            if let Some(num_str) = part.trim().split_whitespace().next() {
                if let Ok(num) = num_str.parse::<usize>() {
                    return num;
                }
            }
        }
    }
    0
}
